Output buffer check in addition to the existing headers_sent() check for Redis session handling.

cyberguroo opened the pull request:
### Summary of Changes
Added additional check for output buffer state before proceeding to dynamically modify session settings via. ini_set since the function `headers_sent()` is proving inadequate for the 'headers already sent check'.

If the above check is skipped the application throws the following warning

`Warning: ini_set(): A session is active. You cannot change the session module's ini settings at this time in /var/www/html/xxxx/libraries/joomla/session/storage/redis.php on line 58`

### Testing Instructions

Try programatically registering a user and then logging them in into the system.

            jimport('joomla.user.helper');
            $user = new JUser;


            if (!$user->guest) {
                jexit();
            }
            //Write to database
            if (!$user->bind($data)) {
                jexit();
            }
            if (!$user->save()) {
                jexit();
            }
            $user_id = $user->id;
            if (!isset($user->id) || empty($user->id)) {
                jexit();
            }



            $options = array();
            $credentials = array();

            $credentials['username'] = $email;
            $credentials['password'] = $tempUser['password'];
            echo "before logging in";
             // this line will throw the aforementioned error.
            $result = JFactory::getApplication()->login($credentials, $options);



### Expected result
The system should not throw the warning since a "headers sent check" is already put in place.

### Actual result
Since output buffering though discouraged, still is allowed in joomla the user has a possibility of having it on, and 'headers_sent()' doesn't register the headers sent to output buffer.

as explained here 
http://www.plus2net.com/php_tutorial/headers_sent.php



### Documentation Changes Required
None

Commit:
Added additional check for output buffer state before proceeding to dynamically modify session settings via. ini_set since the function `headers_sent()` is proving inadequate for the 'headers already sent check'.

If the above check is skipped the application throws the following error 

`Warning: ini_set(): A session is active. You cannot change the session module's ini settings at this time in /var/www/html/xxxx/libraries/joomla/session/storage/redis.php on line 58`

cyberguroo committed:
Update redis.php

Replaced spaces with tabs.

cyberguroo committed:
Update redis.php

cyberguroo committed:
Update redis.php

cyberguroo committed:
Update redis.php

cyberguroo commented:
Need more testing, shall reopen with full solution.

cyberguroo committed:
Update redis.php

Ok, So this problem was two fold, 

1. The headers sent when output_buffer is on.
2. The register method being re called when programatically logging a user in.

Added the respective checks on line 65 and 85 respectively.

Quy commented on libraries/joomla/session/storage/redis.php:
```diff
@@ -82,9 +82,11 @@ public function register()
-				ini_set('session.save_path', $path);
-				ini_set('session.save_handler', 'redis');
+				// Check if session has already been started, depending on the php.ini settings the active status can be either a "2" or "PHP_SESSION_ACTIVE"
+				if ((session_status() !== 2) && (session_status() !== "PHP_SESSION_ACTIVE")) {
```
Please move `{` to its own line.

cyberguroo committed:
Update redis.php

Removed white spaces and corrected indentations.

cyberguroo committed:
Update redis.php

Fixed following error (for < php 5.5 ) by moving the function out of the empty(). 

Fatal error: Can't use function return value in write context

laoneo committed:
Merge branch '3.10-dev' into fix-redis-session-pageheader-check

laoneo commented:
Sorry that it took so long to respond. Can you update the description of this pr in which context your issue occurs? At the moment this looks like a special situation. As Joomla has evolved there might be other ways to implement your use case. In the meantime I'm closing the pr, when updated, please reopen again or make one for Joomla 4. Thanks for your help making Joomla better.
